crates/jiter/src/py_error.rs

```diff
@@ -0,0 +1,210 @@
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+
+use crate::errors::{JsonError, LinePosition};
+
+#[pyclass(extends=PyValueError, module="jiter")]
+#[derive(Debug, Clone)]
+pub struct JsonParseError {
+    json_error: JsonError,
+    path: Vec<PathItem>,
+    position: LinePosition,
+}
+
+impl JsonParseError {
+    pub(crate) fn new_err(py: Python, py_json_error: PythonJsonError, json_data: &[u8]) -> PyErr {
+        let position = py_json_error.error.get_position(json_data);
+        let slf = Self {
+            json_error: py_json_error.error,
+            path: match py_json_error.path {
+                Some(mut v) => {
+                    v.reverse();
+                    v
+                }
+                None => vec![],
+            },
+            position,
+        };
+        match Py::new(py, slf) {
+            Ok(err) => PyErr::from_value_bound(err.into_bound(py).into_any()),
+            Err(err) => err,
+        }
+    }
+}
+
+#[pymethods]
+impl JsonParseError {
+    fn kind(&self) -> &'static str {
+        self.json_error.error_type.kind()
+    }
+
+    fn description(&self) -> String {
+        self.json_error.error_type.to_string()
+    }
+
+    fn path(&self, py: Python) -> PyObject {
+        self.path.to_object(py)
+    }
+
+    fn index(&self) -> usize {
+        self.json_error.index
+    }
+
+    fn line(&self) -> usize {
+        self.position.line
+    }
+
+    fn column(&self) -> usize {
+        self.position.column
+    }
+
+    fn __str__(&self) -> String {
+        format!("{} at {}", self.json_error.error_type, self.position)
+    }
+
+    fn __repr__(&self) -> String {
+        format!("JsonParseError({:?})", self.__str__())
+    }
+}
+
+pub(crate) trait MaybeBuildArrayPath: MaybeBuildPath {
+    fn incr_index(&mut self);
+    fn set_index_path(&self, err: PythonJsonError) -> PythonJsonError;
+}
+
+pub(crate) trait MaybeBuildObjectPath: MaybeBuildPath {
+    fn set_key(&mut self, key: &str);
+
+    fn set_key_path(&self, err: PythonJsonError) -> PythonJsonError;
+}
+
+pub(crate) trait MaybeBuildPath {
```

Comment:
I understand why you went this way (and map exceptions as they propagate through `take_value`), though I wonder if it might be better to have a method on `Parser` to work out the `.path()` given some index, similar to how we have a method to work out file position. We could probably re-use a lot of the `.skip()` implementation to achieve it.

I think that'd be the only way to get json-path-to-errors for iterable parsing without keeping wasted state for the 99.99999% of successful parses.

Reply:
ye, you're right, we should probably try that, partly because it avoids the parse struct becoming more generic.